### PR TITLE
[Merged by Bors] - Example for 2D Frustum Culling

### PR DESCRIPTION
**Pull request by Byteron**
This adds a new project for showing off Frustum Culling.
(Master runs this at sub 1 FPS while with the frustum culling it runs at 144 FPS on my system)

Short clip of the project running:
https://streamable.com/vvzh2u

### CODE REVIEW

File: examples/2d/frustum_sprites.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    assets: Res<AssetServer>,
+    mut materials: ResMut<Assets<ColorMaterial>>,
+) {
+    let tile_size = Vec2::new(64.0, 64.0);
+    let map_size = Vec2::new(320.0, 320.0);
+
+    let half_x = (map_size.x / 2.0) as i32;
+    let half_y = (map_size.y / 2.0) as i32;
+
+    let material = materials.add(assets.load("textures/icon.png").into());
```
Comment (DJMcNab) [inline examples/2d/frustum_sprites.rs:37]: Maybe `sprite` would be a better name for this variable?

File: examples/2d/frustum_sprites.rs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        .run()
+}
+
+fn setup(
+    commands: &mut Commands,
+    assets: Res<AssetServer>,
+    mut materials: ResMut<Assets<ColorMaterial>>,
+) {
+    let tile_size = Vec2::new(64.0, 64.0);
+    let map_size = Vec2::new(320.0, 320.0);
```
Comment (DJMcNab) [inline examples/2d/frustum_sprites.rs:32]: These can both use `Vec2::splat`. Not sure if that is clearer, but it might be better than repeating the constant.

File: examples/2d/frustum_sprites.rs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            let position = Vec2::new(x as f32, y as f32);
+            let translation = (position * tile_size).extend(0.0);
+
+            commands.spawn(SpriteBundle {
+                material: material.clone(),
+                transform: Transform::from_translation(translation),
+                sprite: Sprite {
+                    size: tile_size,
+                    resize_mode: SpriteResizeMode::Manual,
+                },
```
Comment (DJMcNab) [inline examples/2d/frustum_sprites.rs:55]:
```suggestion
                sprite: Sprite::new(tile_size),
```

File: examples/2d/frustum_sprites.rs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    }
+}
+
+pub fn move_camera(
+    time: Res<Time>,
+    sprites: Query<&Sprite>,
+    visible: Query<&Sprite, With<Visible>>,
+    mut query: Query<(&mut Transform, &mut Direction, &mut CameraTimer)>,
+) {
+    for (mut transform, mut direction, mut timer) in query.iter_mut() {
```
Comment (DJMcNab) [inline examples/2d/frustum_sprites.rs:68]: Technically, these should probably be multiple queries/systems since the direction swapping is completely unrelated to swapping the direction

That being said, in this case it's fine.

File: examples/2d/frustum_sprites.rs [hunk 5]
```diff
@@ -0,0 +1,83 @@
+use bevy::{
+    diagnostic::{FrameTimeDiagnosticsPlugin, LogDiagnosticsPlugin},
+    prelude::*,
+};
+
+const CAMERA_SPEED: f32 = 1000.0;
+
+pub struct CameraTimer(Timer);
+pub struct Direction(Vec3);
```
Comment (DJMcNab) [inline examples/2d/frustum_sprites.rs:9]: I wonder if a more circular camera motion is more helpful - that way the sprites are always on screen. In that case, you wouldn't need a timer since you would just move around the circle by `SPEED * time.delta_seconds()` radians.

**Review by DJMcNab — CHANGES_REQUESTED** [review]
Please can you add this example to `examples/README.md`

I'd also rename it to something like `many_sprites` personally, although the `2d` folder is definitely correct.

Additionally, even though there aren't many comments in other examples, a comment about 'this example is used to test performance with large amounts of sprites' and a link to #1492 in that comment would be useful.

A good start though

File: examples/2d/frustum_sprites.rs (hunk 5 above)
Comment (DJMcNab) [reply examples/2d/frustum_sprites.rs:9]: Ok, so looking at your video it looks like if this was performing properly then sprites would always be on the screen - the sprites being offscreen is just a quirt of time.deltra_time being 10 seconds (debug mode 😓 )

**Comment by mockersf** [comment]
While a great example a frustum culling, I don't think this is an example of Bevy. It doesn't show any feature that one may wonder "ok, how do I do this with Bevy, let me check the examples".

Do we want to have "showcase" examples for things like performance?

File: examples/2d/frustum_sprites.rs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+fn main() {
+    App::build()
+        .insert_resource(ReportExecutionOrderAmbiguities)
```
Comment (mockersf) [inline examples/2d/frustum_sprites.rs:13]: To avoid having stuff unrelated to the example, this resource should not be added here

File: examples/2d/frustum_sprites.rs (hunk 1 above)
Comment (mockersf) [reply examples/2d/frustum_sprites.rs:37]: Could you reuse "assets/branding/icon.png" here? This would avoid adding a new asset to the repository

**Review by mockersf — CHANGES_REQUESTED** [review]

**Comment by Byteron** [comment]
Gave it a second pass. Now the camera is moving in circles around the tilemap. :P
Also implemented all your suggestions.

File: examples/2d/many_sprites.rs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+///This example is for performance testing purposes.
+///See https://github.com/bevyengine/bevy/pull/1492
+fn main() {
+    App::build()
+        .add_plugin(LogDiagnosticsPlugin::default())
+        .add_plugin(FrameTimeDiagnosticsPlugin::default())
+        .insert_resource(bevy::log::LogSettings {
+            level: bevy::log::Level::DEBUG,
+            ..Default::default()
+        })
```
Comment (DJMcNab) [inline examples/2d/many_sprites.rs:21]: Why is this required? By default the `INFO` of the `LogDiagnosticPlugin` is displayed properly

```suggestion
```

**Review by DJMcNab — APPROVED** [review]
Looks good :)

The video you have linked in the OP needs updating

Also, as of recently you can directly upload videos to GitHub which are embedded

File: examples/2d/many_sprites.rs [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        position.0 = position.0
+            * Transform::from_translation(Vec3::unit_x() * CAMERA_SPEED * time.delta_seconds());
+        transform.translation = position.0.translation;
+    }
+}
+
+pub fn tick(
+    time: Res<Time>,
+    sprites: Query<&Sprite>,
+    visible: Query<&Sprite, With<Visible>>,
```
[the remark this answers is not in the document]
Comment (Byteron) [reply examples/2d/many_sprites.rs:76]: Right.
It does depend on the Culling PR now, as it needs to use the new marker component that is gonna be introduced.

[the remark this answers is not in the document]
Comment (DJMcNab) [reply examples/2d/many_sprites.rs:76]: Well this example could still be merged, then #1492 would change `Visible` to that marker component.
